### PR TITLE
chore: introduce ginkgo framework for backend E2E testing

### LINKED ISSUE #942: test: clean the temporary data in etcd at the end of test case
the current way: https://github.com/apache/apisix-dashboard/blob/96e3ad1ec87f35d52d75f0a560dc152843e21e23/api/test/e2e/route_test.go#L227-L260

The current way is to completely depend on the developer and bring extra work to the test case maintainer.

we need a common way to clean all of the temporary data in etcd after running the test cases of each file.

### LINKED ISSUE #1136: Adding useful information to backend E2E test cases
# Feature request

## Please describe your feature

- [x] Which test case is failed (would better support error highlighting);
- [x] Which file is this test case in?

cc @idbeta @nic-chen

### PR DESCRIPTION
**Pull request by nic-chen**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches
- [x] Testcase

introduce ginkgo framework to backend E2E testing, for better test result displaying and better practice.

- Related issues
close #1136
close #942

### CODE REVIEW

**Comment by juzhiyuan** [comment]
any update?

**Comment by idbeta** [comment]
Can you provide us with what will be the output when the case fails? Is it more convenient to locate the reason?

File: .github/workflows/backend-e2e-test.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ on:
       - master
 
 jobs:
-  backend-e2e-test:
+  backend-e2e-test-old:
```
Comment (idbeta) [inline .github/workflows/backend-e2e-test.yml:12]: Why not "new"?

File: api/test/e2enew/base/base.go [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+	token string
+
+	UpstreamIp             = "172.16.238.20"
+	APISIXHost             = "http://127.0.0.1:9080"
+	APISIXInternalUrl      = "http://172.16.238.30:9080"
+	APISIXSingleWorkerHost = "http://127.0.0.1:9081"
+	ManagerAPIHost         = "http://127.0.0.1:9000"
+)
+
+func GetToken() string {
```
Comment (idbeta) [inline api/test/e2enew/base/base.go:47]: Here it is best to pass in `username ` and `password ` as parameters.

File: api/test/e2enew/base/base.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+	respBody, _ := ioutil.ReadAll(resp.Body)
+
+	return string(respBody)
+}
+
+func CleanResource(resource string) {
+	resources := GetResourceList(resource)
+	list := gjson.Get(resources, "data.rows").Value().([]interface{})
+	for _, item := range list {
+		route := item.(map[string]interface{})
```
Comment (idbeta) [inline api/test/e2enew/base/base.go:280]: better do not use `route` here.

File: api/test/e2enew/base/base.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+		return token
+	}
+
+	requestBody := []byte(`{
+		"username": "admin",
+		"password": "admin"
+	}`)
+
+	url := ManagerAPIHost + "/apisix/admin/user/login"
+	req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(requestBody))
```
Comment (idbeta) [inline api/test/e2enew/base/base.go:58]: It is recommended to unify the HTTP request method, do not sometimes `HttpGet...`, sometimes `http.NewRequest...`

File: api/test/e2enew/base/base.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+	_, err = cmd.CombinedOutput()
+	if err != nil {
+		fmt.Println("cmd error:", err.Error())
+	}
+	assert.Nil(t, err)
+}
+
+func GetResourceList(resource string) string {
+	t := getTestingHandle()
+	request, _ := http.NewRequest("GET", ManagerAPIHost+"/apisix/admin/"+resource, nil)
```
Comment (idbeta) [inline api/test/e2enew/base/base.go:266]: ditto

File: .github/workflows/backend-e2e-test.yml (hunk 1 above)
Comment (membphis) [reply .github/workflows/backend-e2e-test.yml:12]: this is a bad name

**Review by membphis — COMMENTED** [review]
can we start with a small step?

big PR

File: api/test/e2enew/go.sum [hunk 6]
```diff
@@ -0,0 +1,136 @@
+github.com/ajg/form v1.5.1 h1:t9c7v8JUKu/XxOGBU0yjNpaMloxGEJhUkqFRq0ibGeU=
```
Comment (membphis) [inline api/test/e2enew/go.sum:1]: please confirm the license

**Review by starsz — APPROVED** [review]

File: api/test/e2enew/base/base.go (hunk 4 above)
Comment (nic-chen) [reply api/test/e2enew/base/base.go]: fixed

File: api/test/e2enew/base/base.go (hunk 3 above)
Comment (nic-chen) [reply api/test/e2enew/base/base.go]: fixed.

**Comment by nic-chen** [comment]
@membphis  license check result:
```
github.com/yudai/gojsondiff,https://github.com/yudai/gojsondiff/blob/master/LICENSE,MIT
github.com/nxadm/tail,https://github.com/nxadm/tail/blob/master/LICENSE,MIT
github.com/gavv/httpexpect/v2,https://github.com/gavv/httpexpect/blob/master/v2/LICENSE,MIT
github.com/fsnotify/fsnotify,https://github.com/fsnotify/fsnotify/blob/master/LICENSE,BSD-3-Clause
github.com/imkira/go-interpol,https://github.com/imkira/go-interpol/blob/master/LICENSE,MIT
github.com/klauspost/compress,https://github.com/klauspost/compress/blob/master/LICENSE,BSD-3-Clause
github.com/yalp/jsonpath,https://github.com/yalp/jsonpath/blob/master/LICENSE,BSD-3-Clause
github.com/sergi/go-diff/diffmatchpatch,https://github.com/sergi/go-diff/blob/master/diffmatchpatch/LICENSE,MIT
golang.org/x/net/publicsuffix,Unknown,BSD-3-Clause
github.com/nxadm/tail/ratelimiter,https://github.com/nxadm/tail/blob/master/ratelimiter/Licence,MIT
golang.org/x/sys,Unknown,BSD-3-Clause
github.com/gorilla/websocket,https://github.com/gorilla/websocket/blob/master/LICENSE,BSD-2-Clause
github.com/pmezard/go-difflib/difflib,https://github.com/pmezard/go-difflib/blob/master/difflib/LICENSE,BSD-3-Clause
github.com/xeipuuv/gojsonschema,Unknown,Apache-2.0
github.com/xeipuuv/gojsonreference,Unknown,Apache-2.0
github.com/xeipuuv/gojsonpointer,Unknown,Apache-2.0
moul.io/http2curl,Unknown,Apache-2.0
github.com/fatih/structs,https://github.com/fatih/structs/blob/master/LICENSE,MIT
gopkg.in/yaml.v2,Unknown,Apache-2.0
github.com/onsi/ginkgo/reporters/stenographer/support/go-colorable,https://github.com/onsi/ginkgo/blob/master/reporters/stenographer/support/go-colorable/LICENSE,MIT
github.com/tidwall/gjson,https://github.com/tidwall/gjson/blob/master/LICENSE,MIT
github.com/ajg/form,https://github.com/ajg/form/blob/master/LICENSE,BSD-3-Clause
github.com/yudai/golcs,https://github.com/yudai/golcs/blob/master/LICENSE,MIT
github.com/valyala/fasthttp,https://github.com/valyala/fasthttp/blob/master/LICENSE,MIT
github.com/stretchr/testify,https://github.com/stretchr/testify/blob/master/LICENSE,MIT
github.com/davecgh/go-spew/spew,https://github.com/davecgh/go-spew/blob/master/spew/LICENSE,ISC
github.com/valyala/bytebufferpool,https://github.com/valyala/bytebufferpool/blob/master/LICENSE,MIT
github.com/onsi/ginkgo,https://github.com/onsi/ginkgo/blob/master/LICENSE,MIT
gopkg.in/tomb.v1,Unknown,BSD-3-Clause
github.com/tidwall/pretty,https://github.com/tidwall/pretty/blob/master/LICENSE,MIT
github.com/google/go-querystring/query,https://github.com/google/go-querystring/blob/master/query/LICENSE,BSD-3-Clause
github.com/tidwall/match,https://github.com/tidwall/match/blob/master/LICENSE,MIT
github.com/klauspost/cpuid,https://github.com/klauspost/cpuid/blob/master/LICENSE,MIT
```

**Review by gxthrj — APPROVED** [review]

**Pull request merged**